### ISSUE TITLE
[Feature request] DS virtualpad(add and substract X, Y value automatically)

### ISSUE THREAD

**Issue by InfamousKnight** [TASVideos/BizHawk#1925]
So to be more specific, in the DS virtualpad, you have the X and Y coordinates for the touch screen, and buttons to add and substract them manually, but it would be nice to have another box below it to automatically add and substract on frame advance. So then we don't have to painstakingly add/substract every time we advance a frame.

Sound like a good idea?

**Comment by InfamousKnight** [comment]
Is the title change equivalent to what it was originally?

**Comment by adelikat** [comment]
I like this idea, also it will affect other peripherals too such as zappers for various platforms, and I think it might be a good idea there too.

**Comment by InfamousKnight** [comment]
Yeah, and this should also increase motivation to make ds tases. Which was the whole point of using bizhawk instead of desmume.

**Comment by heckez-sys** [comment]
@InfamousKnight can i have the bios for nds cores plz and thx even if you dont give them
(yes i use plz thx ur when typing)

**Comment by InfamousKnight** [comment]
Thats off topic. Really has nothing to do with what we are discussing here.

**Comment by InfamousKnight** [comment]
I have looked into implementing this myself, but was a little lost on the code. I haven't done any programming since 2014.

Wouldn't this be a good start if I were to help you guys?

**Comment by YoshiRulz** [comment]
It's not the worst idea. The relevant code is the WinForms Control [`VirtualPadTargetScreen`](https://github.com/TASVideos/BizHawk/blob/master/src/BizHawk.Client.EmuHawk/tools/VirtualPads/controls/VirtualPadTargetScreen.cs).

**Comment by InfamousKnight** [comment]
I thought it was here: https://github.com/TASVideos/BizHawk/blob/master/src/BizHawk.Client.EmuHawk/tools/VirtualPads/schema/NdsSchema.cs

I'm more lost with the code you sent than the one I looked at in the first place.

**Comment by YoshiRulz** [comment]
The schemata specify the layouts for each type of controller. This `switch` in a loop creates the actual UI elements from those layouts:
https://github.com/TASVideos/BizHawk/blob/41109d6fc26f89ab82e456863eb01e6adf7f7e20/src/BizHawk.Client.EmuHawk/tools/VirtualPads/VirtualPad.cs#L81-L132

**Comment by InfamousKnight** [comment]
Yeah, I looked through there, I'm still really lost..

I just don't know where I would add the add and subtract x and y
coordinates. And wouldn't we have to tell another schema about that? How
does it know that its for ds virtual pad?

**Comment by zeromus** [comment]
isn't this something you can do with lua?

**Comment by InfamousKnight** [comment]
I'm not sure given how adilikats cross hair lua script doesn't run. Lua
doesn't work on Linux either, so having it on virtual pad would make this
more cross platform. I understand melonds still hasn't been compiled for
Linux, but from the looks of lua support for Linux, they say it may never
happen. Melonds looks like there's some hope left for it.

**Comment by zeromus** [comment]
There's no reason lua can't work on linux. Don't we already support a fully managed lua in nlua+kopilua?

**Comment by RetroEdit** [comment]
@zeromus: YoshiRulz has documented some stuff about Linux compatibility in #1430 and I get the general impression #2260 is thought to be the way forward. But this seems like we're getting off-topic a bit because Lua not being supported in Linux is a more general problem we're already aware of that we need to solve for Linux compatibility progress.

**Comment by zeromus** [comment]
If lua can solve the problem then it's not off topic. We should not add junked up code for something lua should do, just because we can't right now get lua working on linux.

**Comment by RetroEdit** [comment]
I agree we shouldn't add junked up code. That doesn't necessarily mean some amount of feature redundancy is bad. Technically, everything you do in the Virtual Pad could be recoded from scratch in a Lua script, but that doesn't mean we should scrap the Virtual Pad completely and make the user code these interactions from scratch in Lua.

This feature request seems to be specifically about the Virtual Pad, so suggesting Lua doesn't actually address the issue, it only suggests closing the issue and instead using Lua. Which is fine if that's what we want to do, but that's not what I'm hearing from adelikat's comment.

**Comment by adelikat** [comment]
Lua can do things.  But it doesn't man that are TargetPair control couldn't be better, it doesn't do anything particularly useful currently, and could be improved.

**Comment by YoshiRulz** [comment]
Two possible implementations:
1. New checkbox in `VirtualPadTargetScreen` which changes its behaviour: the previous frame's output values are cached, and in relative mode, added to the current pos before output to the `XORWhatever`. May be annoying to click near the origin to get, for example, (2, 1) for 2 px over and 1 px up each frame. Also I think the origin is in the top-left for the DS screen and other zapper-like axis pairs, no idea what to do about that.
2. New checkbox which replaces `VirtualPadTargetScreen` with a wrapper around `VirtualPadAnalogStick`, this wrapper doing the diff of values. (The two controls and the checkbox will also need to be in a parent control.) `VirtualPadAnalogStick` technically also uses absolute coordinates, but I think users' experience with "push the stick less to go slower" from 3D platformers will carry over to make it intuitive. We can limit the range of the stick to -10..10 or thereabouts (maybe configurable).
